### ISSUE TITLE
JavaScript Authenticator Providers not updated automatically on build and SQL error when removing/adding in flow

### ISSUE THREAD

**Issue by Cracky5457** [keycloak/keycloak#20005]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

core

### Describe the bug

If you package a Javascript provider and deploy it to a jar, add the scripts to a flow, and then edit and redeploy the script, re build the server and start, the scripts is still in the old versions

Workaround : Remove the scripts from the flow and add it again is the only fix I found.

Edit : Actually this workaround worked the first time but now it's even worst

If I remove the Script from the flow and if I add it again I got the error

```
2023-04-27 20:11:36,510 WARN  [org.mariadb.jdbc.message.server.ErrorPacket] (executor-thread-4) Error: 1062-23000: Duplicate entry 'script-role-based-access.js' for key 'PRIMARY'
2023-04-27 20:11:36,511 WARN  [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (executor-thread-4) SQL Error: 1062, SQLState: 23000
2023-04-27 20:11:36,511 ERROR [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (executor-thread-4) (conn=542) Duplicate entry 'script-role-based-access.js' for key 'PRIMARY'
```

`script-role-based-access.js' is the name of my scripts here

### Version

21.1.1

### Expected behavior

Should update the script

### Actual behavior

Script already in the flow are not updated

### How to Reproduce?

- Install a Javascript providers (in my case an authenticator)
- Start the server and the script to a flow (in my case Browser flow copy)
- edit the content of the script, rebuild the jar `my-scripts.jar`
- update`providers/my-scripts.jar` and rebuild `bin/kc.sh build --features=scripts`then start

Now if you go in your flow and display the preview of your script, you still see the previous content

I tried to remove/rename the jar without success

Only fix for me was to remove the script from the flow and add it again (which is an easy workaround) but this doesn't always works second time I had

```
2023-04-27 20:11:36,510 WARN  [org.mariadb.jdbc.message.server.ErrorPacket] (executor-thread-4) Error: 1062-23000: Duplicate entry 'script-role-based-access.js' for key 'PRIMARY'
2023-04-27 20:11:36,511 WARN  [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (executor-thread-4) SQL Error: 1062, SQLState: 23000
2023-04-27 20:11:36,511 ERROR [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (executor-thread-4) (conn=542) Duplicate entry 'script-role-based-access.js' for key 'PRIMARY'
```

Fix was 

`DELETE FROM AUTHENTICATOR_CONFIG_ENTRY WHERE AUTHENTICATOR_ID='script-role-based-access.js';`
`DELETE FROM AUTHENTICATOR_CONFIG WHERE ID='script-role-based-access.js';`

**Comment by Cracky5457** [comment]
Maybe already fixed in 21.1.1 by

https://github.com/keycloak/keycloak/issues/19867

**Comment by mposolda** [comment]
I am not able to reproduce with latest Keycloak main. What I did was to test with some testing Keycloak script providers. I've also tried to test with both H2 and MariaDB, but was not able to reproduce.

Can you please add more details about your steps to reproduce and how they differ from mine? What I exactly did was:

1) In Keycloak codebase main, I've deployed script providers:
```
cd testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers-deployment
mvn clean install
cp target/integration-arquillian-testsuite-providers-deployment-999.0.0-SNAPSHOT.jar $KEYCLOAK_HOME/providers/integration-arquillian-testsuite-providers-deployment-999.0.0-SNAPSHOT.jar
```

2) Started Keycloak with command like:
```
./kc.sh start-dev --db=mariadb --db-url=jdbc:mariadb://localhost/keycloak --db-username=keycloak --db-password=keycloak --features=scripts
```

3) Created admin user and login to admin console.

4) Created realm `test`, then created authentication flow with the browser changed to something like:

```
...
Copy of Browser forms ALTERNATIVE
-> Username Password Form REQUIRED
-> script-auth-example-js REQUIRED
```

5) Created user `john` and user `fail`

6) Went to http://localhost:8080/realms/test/account . Login as user `john` -> was OK, Login as user `fail` -> failed (Expected behaviour)

7) Did some change in the script. Updated this line https://github.com/keycloak/keycloak/blob/main/testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers-deployment/src/main/resources/scripts/auth-example.js#L5 to user `john`

8) Re-deployed the script (Step 1 done again)

9) Restarted the server

10) Repeated the step 6. Now I am able to authenticate as user `fail` and not able to authenticate as user `john`. So expected behaviour.

Can you check with latest Keycloak main and/or with latest Keycloak nightly image from https://github.com/keycloak/keycloak/releases/download/nightly/keycloak-999.0.0-SNAPSHOT.tar.gz and doublecheck if the bug is still applied?

Eventually, if you can still see the bug, please add the *detailed* steps to reproduce (especially add where is the difference between mine steps to reproduce and yours). Thanks!

**Comment by raoganeshr** [comment]
> I am not able to reproduce with latest Keycloak main. What I did was to test with some testing Keycloak script providers. I've also tried to test with both H2 and MariaDB, but was not able to reproduce.
> 
> Can you please add more details about your steps to reproduce and how they differ from mine? What I exactly did was:
> 
> 1. In Keycloak codebase main, I've deployed script providers:
> 
> ```
> cd testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers-deployment
> mvn clean install
> cp target/integration-arquillian-testsuite-providers-deployment-999.0.0-SNAPSHOT.jar $KEYCLOAK_HOME/providers/integration-arquillian-testsuite-providers-deployment-999.0.0-SNAPSHOT.jar
> ```
> 
> 2. Started Keycloak with command like:
> 
> ```
> ./kc.sh start-dev --db=mariadb --db-url=jdbc:mariadb://localhost/keycloak --db-username=keycloak --db-password=keycloak --features=scripts
> ```
> 
> 3. Created admin user and login to admin console.
> 4. Created realm `test`, then created authentication flow with the browser changed to something like:
> 
> ```
> ...
> Copy of Browser forms ALTERNATIVE
> -> Username Password Form REQUIRED
> -> script-auth-example-js REQUIRED
> ```
> 
> 5. Created user `john` and user `fail`
> 6. Went to http://localhost:8080/realms/test/account . Login as user `john` -> was OK, Login as user `fail` -> failed (Expected behaviour)
> 7. Did some change in the script. Updated this line https://github.com/keycloak/keycloak/blob/main/testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers-deployment/src/main/resources/scripts/auth-example.js#L5 to user `john`
> 8. Re-deployed the script (Step 1 done again)
> 9. Restarted the server
> 10. Repeated the step 6. Now I am able to authenticate as user `fail` and not able to authenticate as user `john`. So expected behaviour.
> 
> Can you check with latest Keycloak main and/or with latest Keycloak nightly image from https://github.com/keycloak/keycloak/releases/download/nightly/keycloak-999.0.0-SNAPSHOT.tar.gz and doublecheck if the bug is still applied?
> 
> Eventually, if you can still see the bug, please add the _detailed_ steps to reproduce (especially add where is the difference between mine steps to reproduce and yours). Thanks!

Hi Marek,

Are you able to create a step for the script in more than one realm? That used to fail in Keycloak 21.

**Comment by chrisFrodo** [comment]
Hi !

@mposolda 

I think I can add some infos about this error.
Behind this there is a overall issue about primary keys for the custom authenticator scripts in the DB.

In Keycloak 21.1.1, when you look at the DB table "AUTHENTICATOR_CONFIG" the primary key is "ID" with a length of 36 chars (like a UUID).

All built-in authenticators have UUID primary keys, but in the case of custom authenticator scripts the primary key used is the **name** of the script... It means that when you use a script in a flow in the Master realm you cannot use it again in a different realm.

Same thing happens if you want to use twice the same script in the same flow, in the same realm, or even the same step.

**Comment by mposolda** [comment]
@Cracky5457 @chrisFrodo This helps to reproduce, Thanks! We should have it fixed in Keycloak 23 and/or maybe even 22.0.2
